### ISSUE TITLE
SFTP: Add support for ecdsa-sha2-nistp521 keys

### ISSUE THREAD

**Issue by Dennis14e** [duplicati/duplicati#5090]
- [x] I have searched open and closed issues for duplicates.
- [x] I have searched the [forum](https://forum.duplicati.com) for related topics.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.7.1_beta_2023-05-25
- **Operating system**: Docker linuxserver/duplicati:latest
- **Backend**: SFTP (SSH)

## Description

Tried to use ecdsa-sha2-nistp521 key, got error message "Failed to connect: Failed to parse the keyfile, check the key format and passphrase. Error message was openssh key type: ecdsa-sha2-nistp521 is not supported"

## Steps to reproduce

1. Use WebGUI
2. Add backup
3. Configure a new backup
4. Enter some general backup settings
5. Backup destination SFTP (SSH), enter SSH Server options
6. Add advanced option "ssh-keyfile", use an ecdsa-sha2-nistp521 key
7. Get error message on "Test connection"

- **Actual result**:
error message "Failed to connect: Failed to parse the keyfile, check the key format and passphrase. Error message was openssh key type: ecdsa-sha2-nistp521 is not supported"

- **Expected result**:
A successful connection

**Comment by ts678** [comment]
This might be a multipart problem. Searching for "Failed to parse the keyfile" found things like the below. Do they apply here?

[Baaahhh… authentication failing after upgrading ubuntu](https://forum.duplicati.com/t/baaahhh-authentication-failing-after-upgrading-ubuntu/14834) in Duplicati forum.

[Support OpenSSH 8.8 #883](https://github.com/sshnet/SSH.NET/issues/883) in `SSH.NET` Issues -- they are Duplicati's SFTP. Issue solved as of 2023.0.0, but Duplicati needs it.

[Support OpenSSH 8.8 #4615](https://github.com/duplicati/duplicati/issues/4615) is a Duplicati issue, suggesting using their 2023.0.0, however they've recently released 2023.0.1.

After getting past the keyfile parse, the next questions are if they support ecdsa-sha2-nistp521 (they seem to), and on `mono`?

[SFTP backup no longer working on Asustor NAS after patch](https://forum.duplicati.com/t/sftp-backup-no-longer-working-on-asustor-nas-after-patch/16985) was a big chase involving the developers, private builds, and some concern about 2023.0 compatibility. I haven't looked to see if 2023.0.1 looks better or worse generally, compared to 2023.0.0.

SSH can be tough to debug without logs or packets. Do you know your type of server? Best for a look is accessible on Internet.

**Comment by Dennis14e** [comment]
For me it does not look like a server problem, but that the client key is not supported and it does not even try to connect to the server.

It's an Ubuntu server with OpenSSH 8.9 Server.

**Comment by ts678** [comment]
> it does not even try to connect to the server.

I'm assuming you've verified that with something like network tools or pointing to a dead IP address yet having it fail instantly.

Then maybe it's the OpenSSH 8.8 issues that might be fixed in 2023.0.1. I'm not sure where that do-or-not decision is sitting...

**Comment by Dennis14e** [comment]
> I'm assuming you've verified that with something like network tools or pointing to a dead IP address yet having it fail instantly.

Yes, even if I enter a dead IP address as the server, I immediately get the same error message.

**Comment by gpatel-fr** [comment]
@ts678 
>>Then maybe it's the OpenSSH 8.8 issues that might be fixed in 2023.0.1

no it's not. 
This is most probably a Mono problem and as such never will be fixed as long as Duplicati is based on Mono.

So this is going to be a WontFix because going Net.xx (xx > 8) is not for tomorrow.

This said, I think that most modern configured OpenSSH server that do not support ssh-rsa key by default (that can be fixed at the server level of course by setting PubkeyAcceptedAlgorithms ssh-rsa) DO support ed25519 keys out of the box, that are working fine with Duplicati (even with current very old SSH.NET).

**Comment by ts678** [comment]
It looks like the new SSH.NET will also add rsa-sha2-512 rsa-sha2-256.
The problem with ecdsa anything is apparently that `mono` can't do it...

**Comment by ts678** [comment]
> no it's not going to be fixed by 2023.0.1

Since that was said after [update sshnet to 2023.0.1 #5101](https://github.com/duplicati/duplicati/pull/5101) pull request, I'll assume it was tested.
Trying to work out what SSH.NET is up to by reading their source code is a bit too tough for me.

> This is most probably a Mono problem

Indicated by the SSH.NET issues, and if it was now worked around, I'd expect them to close them.
They seem pretty good at that, though not so good at documenting platform specific limitations.

> https://github.com/mono/mono/blob/main/mcs/class/referencesource/System.Core/System/Security/Cryptography/ECDsaCng.cs

was pointed to in one isssue. Note the `NotImplementedException`.

> https://github.com/mono/mono/blob/main/mcs/class/System.Core/System.Security.Cryptography/README.CNG

explains the `Cng` classes as wrappers around the native Windows libraries, with some exceptions such as ECDsaCng.

[Native Library Loading in .NET 5](https://www.mono-project.com/news/2020/08/24/native-loader-net5/) gave some hope about "numerous changes in the available APIs",
but it doesn't seem released, and as .NET 5 was released soon after, .NET 5-mono might never be.
[Introducing .NET 5](https://devblogs.microsoft.com/dotnet/introducing-net-5/) explains what Microsoft sees as the path forward for their cross-platform work.

> never will be fixed as long as Duplicati is based on Mono.

It may be fixed in SSH.NET by using a different cryptography provider, but I don't know if it will be.

So what can user do? Three alternative algorithms got mentioned (not including the ancient ones).
If ecdsa-sha2-nistp521 is the only acceptable one somehow, see if `rclone` can do it, then use the
[Rclone](https://duplicati.readthedocs.io/en/latest/05-storage-providers/#rclone) storage provider in Duplicati to use rclone.

**Comment by gpatel-fr** [comment]
>>I'll assume it was tested.

correct. It will fix some things but not this one.

>>Three alternative algorithms got mentioned

As I said, ed25519 is a good solution for this issue.

**Comment by ts678** [comment]
If anyone is willing to cautiously test a Canary [release](https://github.com/duplicati/duplicati/releases), `mono` has been replaced by .NET 8, so ecdsa-sha2-nistp521 might work.

**Comment by kenkendk** [comment]
Tested this today and it works fine (at least with SSH.NET 2025.0.0)
